### PR TITLE
Check metrics exists in list of metrics include labels submatch tooling.

### PR DESCRIPTION
**Pull request by ppalucki**
two functions:
- `assert_metric` - metrics with matching labels existing in given list and value optionally equals (used for testing for some metrics in `storage.store` calls)
- `assert_subdict` - on dict is a subdict of another in recursive manner (used for testing that `detect`/`allocate` got right data

### CODE REVIEW

**Comment by felidadae** [comment]
Like the idea.

File: tests/test_testing.py [hunk 1]
```diff
@@ -21,3 +37,70 @@ def test_create_open_mock_manually():
...
+    (dict(x=1), dict(), True),  # ditto,
+    (dict(), dict(x=1), False),  # Non empty query doesn't match empty dict,
+    (dict(x=2), dict(x=1), False),  # no when value is different
+    (dict(x=1), dict(x=1), True),  # ok for the same
+    (dict(x=1, y=2), dict(x=1), True),  # additional parameters doesn't interfere
+    (dict(x=1, y=2), dict(x=1, y=2), True),  # two params exact match
+    (dict(x=1, y=1, z=3), dict(x=1, y=2), False),  # two keys mismatch
+    (dict(x=1, y=2, z=3), dict(x=1, y=2), True),  # two vs three keys match
+])
+def test_match_keys(got_dict, expected_subdict, expected_match):
```
Comment (iwankgb) [inline tests/test_testing.py]: This is not about keys; rename the function, please.

File: owca/testing.py [hunk 2]
```diff
@@ -109,3 +109,59 @@ def task(cgroup_path, labels=None, resources=None):
...
+
+def assert_subdict(got_dict: dict, expected_subdict: dict):
+    """Assert that one dict is a subset of another dict in recursive manner.
+    """
+    for expected_key, expected_value in expected_subdict.items():
+        if expected_key not in got_dict:
+            raise AssertionError('key %r not found in %r' % (expected_key, got_dict))
+        got_value = got_dict[expected_key]
+        if isinstance(expected_value, dict):
+            # for dict use subsets
```
Comment (iwankgb) [inline owca/testing.py]: I can't understand your comment. Could you rephrase it?

File: owca/testing.py [hunk 3]
```diff
@@ -109,3 +109,59 @@ def task(cgroup_path, labels=None, resources=None):
...
+            if not isinstance(got_value, dict):
+                raise AssertionError('expected dict type at %r key, got %r' % (
+                    expected_key, type(got_value)))
+            assert_subdict(got_value, expected_value)
+        else:
+            # for other types check ordinary equality operator
```
Comment (iwankgb) [inline owca/testing.py]: We don't care about lists, do we?

File: owca/testing.py [hunk 4]
```diff
@@ -109,3 +109,59 @@ def task(cgroup_path, labels=None, resources=None):
...
+                'value differs got=%r expected=%r at key=%r' % (
+                    got_value, expected_value, expected_key)
+
+
+def _is_dict_match(got_dict: dict, expected_subdict: dict):
+    """Match values and keys from dict (non recursive)."""
+    for expected_key, expected_value in expected_subdict.items():
+        if expected_key not in got_dict:
+            return False
+        else:
```
Comment (iwankgb) [inline owca/testing.py]:
```suggestion
        else if got_dict[expected_key] != expected_value:
```

File: owca/testing.py [hunk 5]
```diff
@@ -109,3 +109,59 @@ def task(cgroup_path, labels=None, resources=None):
...
+            if got_dict[expected_key] != expected_value:
```
Comment (iwankgb) [inline owca/testing.py]:
```suggestion
```

File: owca/testing.py [hunk 6]
```diff
@@ -109,3 +109,59 @@ def task(cgroup_path, labels=None, resources=None):
...
+                return False
```
Comment (iwankgb) [inline owca/testing.py]:
```suggestion
            return False
```

File: tests/test_testing.py [hunk 7]
```diff
@@ -21,3 +37,70 @@ def test_create_open_mock_manually():
...
+        (dict(x=1, y=dict(z=1)), dict(x=1, y=dict()), None),
+        (dict(x=1, y=dict(z=1)), dict(x=1, y=dict(z=1)), None),
+        (dict(x=1, y=dict(z=1)), dict(y=dict(z=1)), None),
+        # some deep value differs
+        (dict(x=1, y=dict(z=1)), dict(x=1, y=dict(z=2)), "value differs"),
+        (dict(x=1, y=dict(z=1)), dict(y=dict(not_existient_key=2)), "not found"),
+    ])
+def test_assert_subdict(got_dict, expected_subdict, exception_message):
+    if exception_message is not None:
+        with pytest.raises(AssertionError, match=exception_message):
```
Comment (iwankgb) [inline tests/test_testing.py]: Could you extend `exception_message` a bit (to make sure that the exception is raised because of particular value). String "value differs" is very generic and will match a lot of cases, not necessarily the one that we want to match.

File: tests/test_testing.py [hunk 8]
```diff
@@ -21,3 +37,70 @@ def test_create_open_mock_manually():
...
+    'got_dict, expected_subdict, exception_message', [
+        # propery empty or simple dicts
+        (dict(), dict(), None),
+        (dict(x=1), dict(), None),
+        (dict(x=1), dict(), None),
+        (dict(x=1, y=1), dict(x=1), None),
+        # invalid flat dicts
+        (dict(x=1, y=1), dict(x=1, z=2), "key 'z' not found"),
+        (dict(x=1, y=1), dict(x=1, y=2), "value differs"),
+        (dict(x=1, y=1), dict(x=1, y=2), "value differs"),
```
Comment (iwankgb) [inline tests/test_testing.py]:
```suggestion
```
(identical to the case above).

File: owca/testing.py (hunk 3 above)
Comment (ppalucki) [reply owca/testing.py]: no we don't - I don't not want implement and we don't need to use list to check a list is contained by another list (assert_subdict is only for dicts)

File: tests/test_testing.py (hunk 7 above)
Comment (ppalucki) [reply tests/test_testing.py]: the real exception message is "verbose" and tells you everything you need, eg.
'value differs got=%r with expected=%r at key=%r' 
only the match is simple because, because it is regexp and you need to escape all the special characters and makes the test unreadable

File: tests/test_testing.py (hunk 1 above)
Comment (ppalucki) [reply tests/test_testing.py]: fixed - not it is test_is_dict_match

File: owca/testing.py (hunk 2 above)
Comment (ppalucki) [reply owca/testing.py]: Rephrased to:
```If expected value is a dict, call assert_subdict recursively.```

File: owca/testing.py (hunk 4 above)
Comment (ppalucki) [reply owca/testing.py]: fixed - removed else entirely (because `return statement` already ends processing)

File: owca/testing.py (hunk 5 above)
Comment (ppalucki) [reply owca/testing.py]: fixed, explanation above

File: owca/testing.py (hunk 6 above)
Comment (ppalucki) [reply owca/testing.py]: fixed, explanation above

File: tests/test_testing.py (hunk 7 above)
Comment (ppalucki) [reply tests/test_testing.py]: fixed - now comment is much better explaining what was expected and what we got (and where)

File: tests/test_testing.py (hunk 8 above)
Comment (ppalucki) [reply tests/test_testing.py]: fixed
